plugin/pkg/rqdata/rqdata.go

```diff
@@ -0,0 +1,220 @@
+package rqdata
+
+import (
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+type requestFunc func(state request.Request) string
+
+// Mapping define the mapping between 'name' of data and the way to extract that data from the Request
+// it also defines what will be the empty value returned if the data behind the name is empty.
+// it is pretty static, and you should need to instantiate only once
+type Mapping struct {
+	replacements map[string]requestFunc
+	emptyValue   string
```

Comment:
This emptyValue is odd? 1) why is `""` not enough  to signal "not found"

Reply:
Because I plan this tool to be shared with Log. And Log replace an empty value with "-".